File: Custom_Comparators_and_Coordinate_Compression/Silver/P3_2019-Mountain_View/P3_2019-Mountain_View.py
# Custom Comparator Sorting approach
# T: O(nlogn), M: O(n), where n is the number of mountains

class Mountain:
	def __init__(self, x_init: int, x_final: int) -> None:
		self.x_init = x_init
		self.x_final = x_final

with open('mountains.in', 'r') as f:
	# Read first line: n = number of mountains
	n = int(f.readline())
	# For the next n lines calculate the xi, xf coordinates of each mountain
	mountains = []
	for ln in range(n):
		x, y = map(int, f.readline().split())
		mountains.append(Mountain(x-y, x+y))

# sort by first appearing mountains and biggest mountains
mountains.sort(key=lambda m: (m.x_init, -m.x_final))

visable = 1
prev_biggest = mountains[0]

for mountain in mountains:
	if mountain.x_final > prev_biggest.x_final:
		visable += 1
		prev_biggest = mountain

print(visable, file=open('mountains.out', 'w'))

'''
# Custom Comparators and Coordinate Compression approach
# T: O(nlogn), M: O(n), where n is the number of mountains

class Mountain:
	def __init__(self, x_init: int, x_final: int) -> None:
		self.x_init = x_init
		self.x_final = x_final

with open('mountains.in', 'r') as f:
	# Read first line: n = number of mountains
	n = int(f.readline())
	# For the next n lines calculate the xi, xf coordinates of each mountain
	mountains = {}
	for ln in range(n):
		x, y = map(int, f.readline().split())
		if x-y not in mountains:
			mountains[x-y] = Mountain(x-y, x+y)
		else:	# forget about the smaller mountain (compress data)
			xf = max(mountains[x-y].x_final, x+y)
			mountains[x-y] = Mountain(x-y, xf)

# sort mountains by first appearing mountains
mountains = list(mountains.values())
mountains.sort(key=lambda m: m.x_init)

visable = 1
prev_biggest = mountains[0]

for mountain in mountains:
	if mountain.x_final > prev_biggest.x_final:
		visable += 1
		prev_biggest = mountain

print(visable, file=open('mountains.out', 'w'))
'''

File: Arrays_and_Hashing/Custom_Comparators_and_Coordinate_Compression/Silver/P3_2019-Mountain_View/P3_2019-Mountain_View.cpp
// Custom Comparator Sorting approach
// T: O(nlogn), M: O(n), where n is the number of mountains

//#include <bits/stdc++.h>
#include <fstream>
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

int main() {
	// Read input file
	freopen("mountains.in", "r", stdin);
	// Read first line: n = number of mountains
	int n;
	cin >> n;
	// For the next n lines calculate the xi, xf coordinates of each mountain
	struct Mountain { int xInit, xFinal; };
	vector<Mountain> mountains(n);
	for (int ln = 0; ln < n; ln++) {
		int x, y;
		cin >> x >> y;
		mountains[ln].xInit = x-y, mountains[ln].xFinal = x+y;
	}

	// sort by first appearing mountains and biggest mountains
	sort(mountains.begin(), mountains.end(), 
		[](Mountain &a, Mountain &b) {
			if (a.xInit != b.xInit)
				return a.xInit < b.xInit;
			return a.xFinal > b.xFinal;
		});
	
	// Count how many visible peaks Bessie can see
	int visible = 1;
	Mountain prevBiggest = mountains[0];
	for (Mountain mountain : mountains) {
		if (mountain.xFinal > prevBiggest.xFinal) {
			++visible;
			prevBiggest = mountain;
		}
	}

	// Write amount of visible peaks to output file
	freopen("mountains.out", "w", stdout);
	cout << visible << endl;
}

/*
// Custom Comparator and Coordinate Compression approach
// T: O(nlogn), M: O(n), where n is the number of mountains

#include <bits/stdc++.h>
// #include <fstream>
// #include <iostream>
// #include <unordered_map>
// #include <vector>
// #include <algorithm>

using namespace std;

int main() {
	// Read input file
	freopen("mountains.in", "r", stdin);
	// Read first line: n = number of mountains
	int n;
	cin >> n;
	// For the next n lines calculate the xi, xf coordinates of each mountain
	struct Mountain { int xInit, xFinal; };
	unordered_map<int, int> seenXInit;
	vector<Mountain> mountains;
	for (int ln = 0; ln < n; ln++) {
		int x, y;
		cin >> x >> y;
		if (seenXInit.find(x-y) == seenXInit.end()) {
			seenXInit[x-y] = mountains.size();
			mountains.push_back({x-y, x+y});
		}
		// Forget about the smaller mountain (compress data)
		else {
			int prevIdx = seenXInit[x-y];
			mountains[prevIdx].xFinal = max(mountains[prevIdx].xFinal, x+y);
		}
	}

	// sort by first appearing mountains and biggest mountains
	sort(mountains.begin(), mountains.end(), 
		[](Mountain &a, Mountain &b) {
			if (a.xInit != b.xInit)
				return a.xInit < b.xInit;
			return a.xFinal > b.xFinal;
		});
	
	// Count how many visible peaks Bessie can see
	int visible = 1;
	Mountain prevBiggest = mountains[0];
	for (Mountain mountain : mountains) {
		if (mountain.xFinal > prevBiggest.xFinal) {
			++visible;
			prevBiggest = mountain;
		}
	}

	// Write amount of visible peaks to output file
	freopen("mountains.out", "w", stdout);
	cout << visible << endl;
}
*/